cmd/shfmt/main.go

```diff
@@ -155,27 +155,31 @@ Utilities:
 	} else if f, ok := out.(*os.File); ok && terminal.IsTerminal(int(f.Fd())) {
 		color = true
 	}
+	args := flag.Args()
 	if flag.NArg() == 0 {
-		if err := formatStdin(); err != nil {
-			if err != errChangedWithDiff {
-				fmt.Fprintln(os.Stderr, err)
-			}
-			return 1
-		}
-		return 0
-	}
-	if *toJSON {
-		fmt.Fprintln(os.Stderr, "-tojson can only be used with stdin/out")
-		return 1
+		args = []string{"-"}
 	}
 	status := 0
-	for _, path := range flag.Args() {
-		walk(path, func(err error) {
-			if err != errChangedWithDiff {
-				fmt.Fprintln(os.Stderr, err)
+	for _, path := range args {
+		if path == "-" {
```

Comment:
as per my comment above, don't do this in a loop; you can just do:

```
if len(args) == 0 || (len(args) == 1 && args[0] == "-") {
    // stdin
    return ...
}
// file loop
```

Otherwise, we have to think about difficult scenarios like `shfmt - -` or `shfmt - /dev/stdin`. Do we want the tool to block forever? Error in some way? I don't think it's worth keeping those as options.

Reply:
OK, I use that logic now but just went back to your original code for most part, since larger changes are not necessary in that case.